### ISSUE TITLE
Not a reference at C:/perl/perl/site/lib/Queue/DBI.pm line 956.

### ISSUE THREAD

**Issue by niceperl** [guillaumeaubert/Queue-DBI#3]
This code (Windows 7, Strawberry Perl 5.20, Queue::DBI 2.6.1):

    use common::sense;
    use DBI;
    use Queue::DBI;
    use Queue::DBI::Admin;
    
    my $dbh = DBI->connect("dbi:SQLite:dbname=queue.sqlite","","") or die $!;
    my $queue_name = 'my_queue';
    my $queues_admin = Queue::DBI::Admin->new(database_handle => $dbh);
    $queues_admin->create_tables(drop_if_exist => 1);
    my $queue = $queues_admin->create_queue( $queue_name );

    my $queue = Queue::DBI->new(
        'queue_name'      => $queue_name,
        'database_handle' => $dbh,
        'cleanup_timeout' => 3600,
        'verbose'         => 1,
    );

    my $data = {
        block1 => 141592653,
        block2 => 589793238,
        block3 => 462643383,
    };
 
    $queue->enqueue( $data );

shows this:

    Entering cleanup(). at test.pl line 14.
    Retrieving data. at test.pl line 14.
    Creating new Queue::DBI::Element objects. at test.pl line 14.
    Found 0 orphaned element(s). at test.pl line 14.
    Leaving cleanup(). at test.pl line 14.
    Entering enqueue(). at test.pl line 27.
    Element inserted, leaving enqueue(). at test.pl line 27.
    Not a reference at C:/perl/perl/site/lib/Queue/DBI.pm line 956.

I am testing the `enqueue` method

**Comment by guillaumeaubert** [comment]
I just tested on the same configuration without issues:

```
C:\Users\Guillaume\Desktop>ver

Microsoft Windows [Version 6.1.7601]

C:\Users\Guillaume\Desktop>systeminfo | findstr /B /C:"OS Name" /C:"OS Version"
OS Name:                   Microsoft Windows 7 Professional
OS Version:                6.1.7601 Service Pack 1 Build 7601

C:\Users\Guillaume\Desktop>perl -v

This is perl 5, version 20, subversion 2 (v5.20.2) built for MSWin32-x64-multi-thread

Copyright 1987-2015, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

C:\Users\Guillaume\Desktop>perl -MQueue::DBI -le "print $Queue::DBI::VERSION"
2.6.1

C:\Users\Guillaume\Desktop>perl test.pl
Entering cleanup(). at test.pl line 14.
Retrieving data. at test.pl line 14.
Creating new Queue::DBI::Element objects. at test.pl line 14.
Found 0 orphaned element(s). at test.pl line 14.
Leaving cleanup(). at test.pl line 14.
Entering enqueue(). at test.pl line 27.
Element inserted, leaving enqueue(). at test.pl line 27.

C:\Users\Guillaume\Desktop>perl -MStorable -le "print Storable->VERSION"
2.51

C:\Users\Guillaume\Desktop>perl -MMIME::Base64 -le "print MIME::Base64->VERSION"
3.15
```

Some follow-up questions to try to figure out what's happening:

1. Are you using the 32 bits or 64 bits version of Strawberry Perl?
1. Can you do a `print Dumper( $queue );` before and after the last line?
2. Can you check the version of storable with `perl -MStorable -le "print Storable->VERSION"`?
3. Can you check the version of MIME::Base64 with `perl -MMIME::Base64 -le "print MIME::Base64->VERSION"`?

Thank you!

**Comment by niceperl** [comment]
** Question #1 **

    perl -v
    This is perl 5, version 20, subversion 1 (v5.20.1) built for MSWin32-x64-multi-thread

** Question #2 **

    $VAR1 = bless( {
                 'dbh' => bless( {}, 'DBI::db' ),
                 'table_names' => {
                                    'queue_elements' => undef,
                                    'queues' => undef
                                  },
                 'serializer' => {
                                   'thaw' => undef,
                                   'freeze' => undef
                                 },
                 'max_requeue_count' => undef,
                 'lifetime' => undef,
                 'queue_name' => 'my_queue',
                 'queue_id' => 1,
                 'verbose' => 1
               }, 'Queue::DBI' );
    Entering enqueue(). at test.pl line 28.
    Element inserted, leaving enqueue(). at test.pl line 28.
    $VAR1 = bless( {
                 'dbh' => bless( {}, 'DBI::db' ),
                 'table_names' => {
                                    'queue_elements' => undef,
                                    'queues' => undef
                                  },
                 'max_requeue_count' => undef,
                 'max_id' => undef,
                 'serializer' => {
                                   'thaw' => undef,
                                   'freeze' => undef
                                 },
                 'lifetime' => undef,
                 'verbose' => 1,
                 'queue_id' => 1,
                 'queue_name' => 'my_queue'
               }, 'Queue::DBI' );


** Question #3 **

    perl -MStorable -le "print Storable->VERSION"
    2.51

** Question #4 **

    perl -MMIME::Base64 -le "print MIME::Base64->VERSION"
    3.14

Hope it helps

**Comment by niceperl** [comment]
I have discovered what happened. In my first test, I had enqueued a simple string:

my $data = "this is a test";

With that value, the script shows the message I posted:

    Not a reference at C:/perl/perl/site/lib/Queue/DBI.pm line 956.

But if I use a hash, there is no problem. Example:

    my $data = { 'info' => 'this is a test' };

The ´enqueue´ method documentation says: *The data passed can be a scalar or a reference to a complex data structure. There is no limitation on the type of data that can be stored as it is serialized for storage in the database.*

**Comment by guillaumeaubert** [comment]
Nice catch!

My first thought was to just have `freeze()` return scalars as is. But since a serialized data structure and a scalar are indistinguishable, `thaw()` wouldn't know whether to deserialize a data structure or return the string as is.

Forcing scalars into references of scalars for the purpose of serialization on `freeze()` wouldn't work either. `thaw()` would not be able to tell what was a reference to a scalar originally vs what had been forced into one and should be returned as a scalar.

These two problems could be solved by adding extra information to determine whether a stored string is a scalar or a serialized reference.

**Solution 1**

*Add a column to the elements table, to indicate whether a stored string is a scalar or a serialized reference.*

While this is a clean solution from a data standpoint, it would be a big change for current users, so I'm inclined to stay away from this one.

**Solution 2**

*Add a prefix to the stored string, to indicate whether a stored string is a scalar or a serialized reference.*

This would requires a bit of work to maintain backward compatibility and make sure that when users upgrade, old stored data can still be read.  Old data without a prefix would just be assumed to be a serialized data structure, since there's no way to feed scalars right now. But overall, it should be pretty easy to implement and not disruptive for current users, even if it's not necessarily the prettiest solution.

A side note here: for storage purposes and to maintain backward compatibility, I think we would still have to base64-encode the scalar in both cases, even if we're not freezing it. Otherwise, we may have interesting edge cases based on the collation of the user's database.

**Solution 3**

*Change the documentation to indicate that scalars aren't supported.*

Definitely the easiest and cleanest solution, but it is less satisfying from a user standpoint as it forces you to remember to do things like `enqueue( \$string )`.

What are your thoughts?

**Comment by niceperl** [comment]
I vote for Solution 3. Also improve the Synopsis section, and define a valid `$data` (now it is unassigned, and that was the way I took to write a wrong code...). 
Perhaps the enqueue method may raise an exception if a plain scalar (string or number) was passed. I don't know if that maybe dangerous with existing running code

**Comment by guillaumeaubert** [comment]
Sorry for the delay!

I agree, creating a hash on the fly is a neat solution, but backward-compatibility would be an issue. So I went ahead and implemented the third solution, as well as improved both the synopsis and the documentation of enqueue() as you recommended.

Would you mind taking a look at the commits and letting me know if you see anything else that should be improved wrt enqueue() before I merge the changes and create a release?

**Comment by guillaumeaubert** [comment]
Released in version 2.6.2. Thank you, @niceperl !
